### PR TITLE
DCA adapter

### PR DESCRIPTION
**Pull request by maybenotilya**

### CODE REVIEW

File: DCA/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+COPY data data
+
+COPY adapter_base.py adapter_base.py
+
+COPY adapter.py adapter.py
+
+COPY run.py run.py
+
+RUN mkdir input
```
Comment (kirill-ivanov-a) [inline DCA/Dockerfile]: This needs to be taken to the algorithm level. When mounting a directory, a corresponding directory is automatically created

File: DCA/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
+FROM python:3.10
+
+COPY requirements.txt requirements.txt
+
+RUN python3 -m pip install --no-cache-dir -r requirements.txt
+
+WORKDIR /DCA
+
+COPY weights weights
```
Comment (kirill-ivanov-a) [inline DCA/Dockerfile]: Why not copy all the content at once? `COPY . .`

File: DCA/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# DCA
+This folder contains adapter and Dockerfile for [DCA](https://github.com/Luffy03/DCA) algorithm
+
+## Building Docker Image
+1. Download weights from [here](https://drive.google.com/drive/folders/1oenWpYADqd-tTx7JeDQknxRNd3mgW2kQ) and put them into ./weights folder
```
Comment (kirill-ivanov-a) [inline DCA/README.md]: This must be put into docker, all dependencies must be in docker. See [example](https://github.com/prime-slam/line-detection-association-dockers/blob/main/detectors/LETR/Dockerfile) of model download in docker

File: DCA/adapter.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import numpy as np
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py:15]: Let's have nice sorting of imports (first -- external libs import, then external from, then local imports) and separation lines between. See example https://github.com/prime-slam/line-detection-association-dockers/blob/main/detectors/F-Clip/adapter.py

File: DCA/adapter.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+from PIL import Image
+import argparse
+from pathlib import Path
+
+from adapter_base import AdapterBase
+from utils.tools import *
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: [Why import star in Python is a bad idea](https://www.geeksforgeeks.org/why-import-star-in-python-is-a-bad-idea/)

File: DCA/adapter.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+from utils.my_tools import *
+from module.Encoder import Deeplabv2
+from albumentations import Normalize
+from albumentations import *
+import ever as er
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: I don't think you should use this alias

File: DCA/adapter.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+def max_power_of_2(num : int):
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: this function should be in another module, for example in `utils`

File: DCA/adapter.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    p: int = 2
+    while p * 2 <= num:
+        p *= 2
+    return p
+
+
+class DcaAdapter(AdapterBase):
+    def __init__(self, size=1024, factor=2, mode="Urban"):
+        self.size = size
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: size of what?

File: DCA/adapter.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        self.factor = factor
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: What is the factor?

File: DCA/adapter.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        self.mode = mode
+
+    def set_size(self, size):
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: it's worth looking at the `@property` decorator

File: DCA/adapter.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        image = image.convert("RGB")
+        image = image.resize((self.size, self.size))
+        image = np.array(image)
+        print(image.shape)
+        mask = np.zeros([512, 512])
+        transformer = Compose([
+            Normalize(mean=(73.53223948, 80.01710095, 74.59297778),
+                    std=(41.5113661, 35.66528876, 33.75830885),
+                    max_pixel_value=1, always_apply=True),
+            # Normalize(mean=(123.675, 116.28, 103.53),
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: [Why commented code is bad](https://bytedev.medium.com/code-comment-anti-patterns-and-why-the-comment-you-just-wrote-is-probably-not-needed-919a92cf6758#:~:text=Avoid%20commenting%20out%20code%20and,to%20another%20team%20member's%20work.)

File: DCA/adapter.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        # image - RGB
+        image = self._transform_image(image)
+        model = self._build_model()
+        raw_predictions = self._process(model, image)
+        predictions = self._postprocess_predictions(raw_predictions)
+        return predictions
+
+    # requires MxMx3 image
+    def _transform_image(self, image: np.ndarray):
+        image = Image.fromarray(image)
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: Why do you need to additionally convert the np array into an PIL Image object? `cv2.resize`?

File: DCA/adapter.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        image = image.convert("RGB")
+        image = image.resize((self.size, self.size))
+        image = np.array(image)
+        print(image.shape)
+        mask = np.zeros([512, 512])
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: Why 512? What is this number? It's worth creating an expressive constant or adding configurability

File: DCA/adapter.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        transformer = Compose([
+            Normalize(mean=(73.53223948, 80.01710095, 74.59297778),
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: [Why magic numbers are bad](https://stackoverflow.com/questions/47882/what-are-magic-numbers-and-why-do-some-consider-them-bad)

File: DCA/adapter.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+                    std=(41.5113661, 35.66528876, 33.75830885),
+                    max_pixel_value=1, always_apply=True),
+            # Normalize(mean=(123.675, 116.28, 103.53),
+            #           std=(58.395, 57.12, 57.375),
+            #           max_pixel_value=1, always_apply=True),
+            er.preprocess.albu.ToTensor()
+        ], is_check_shapes=False)
+        blob = transformer(image=image, mask=mask)
+        image = blob['image'].to(torch.device('cuda' if torch.cuda.is_available() else 'cpu'))
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: What if I want to run on the CPU, although the GPU is available?

File: DCA/adapter.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+            use_ppm=True,
+            ppm=dict(
+                num_classes=7,
+                use_aux=False,
+                fc_dim=2048,
+            ),  
+            inchannels=2048,
+            num_classes=7
+        )).to('cuda' if torch.cuda.is_available() else 'cpu')
+        ckpt_path = {"Urban" : "./weights/URBAN_0.4635.pth",
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: Avoid hardcoded paths

Comment (kirill-ivanov-a) [reply DCA/adapter.py]: Use `pathlib` for paths

File: DCA/adapter_base.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+from pathlib import Path
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-i", "--input", type=Path, required=True, help="Input images path.")
+    parser.add_argument("-o", "--output", type=Path, required=True, help="Path where to save resulting masks.")
+    return parser.parse_args()
+
+
+# Adapter accepts image as numpy array and returns mask as image in numpy array
```
Comment (kirill-ivanov-a) [inline DCA/adapter_base.py]: This should be included in the docstring of the method or class

File: DCA/adapter_base.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
+import numpy as np
```
Comment (kirill-ivanov-a) [inline DCA/adapter_base.py]: Lets sort imports

File: DCA/adapter_base.py [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+from abc import ABC, abstractmethod
+import argparse
+from pathlib import Path
+
+def get_args():
```
Comment (kirill-ivanov-a) [inline DCA/adapter_base.py]: This function should be placed in a separate module

File: DCA/colors.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
+from PIL import Image
+import os
+import argparse
+from pathlib import  Path
+
+
+def get_args():
```
Comment (kirill-ivanov-a) [inline DCA/colors.py]: I've already seen this somewhere :)

File: DCA/colors.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
+from PIL import Image
+import os
+import argparse
+from pathlib import  Path
```
Comment (kirill-ivanov-a) [inline DCA/colors.py]: To avoid manually aligning all the spaces, it is better to use a linter. There are two spaces here

File: DCA/colors.py [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+        (3, 3, 3) : (0, 0, 255), # water
+        (4, 4, 4) : (128, 0, 128), # barren
+        (5, 5, 5) : (0, 128, 0), # forest
+        (6, 6, 6) : (141, 85, 36), # agriculture
+    }
+
+    args = get_args()
+    args.output.mkdir(exist_ok=True, parents=True)
+    for fname in os.listdir(args.input):
+        img = Image.open(str(args.input) + f"/{fname}").convert("RGB")
```
Comment (kirill-ivanov-a) [inline DCA/colors.py]: Why make an argument of type Path and then convert it to a string? :)

File: DCA/colors.py [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-i", "--input", type=Path, required=True, help="Input images path.")
+    parser.add_argument("-o", "--output", type=Path, required=True, help="Output images path.")
+    return parser.parse_args()
+
+def color_image():
+    colors = {
```
Comment (kirill-ivanov-a) [inline DCA/colors.py]: It is worth making a config for the markup configuration. What does tuple key mean?

File: DCA/data/nj.py [hunk 24]
```diff
@@ -0,0 +1,198 @@
+import random
```
Comment (kirill-ivanov-a) [inline DCA/data/nj.py]: Let's upload the authors' implementation as a separate request, so that it is clear what to review

**Review by kirill-ivanov-a — CHANGES_REQUESTED** [review]

File: DCA/adapter.py (hunk 13 above)
Comment (maybenotilya) [reply DCA/adapter.py]: It turned out mask does nothing at all :)

File: DCA/adapter.py (hunk 14 above)
Comment (maybenotilya) [reply DCA/adapter.py]: It was in author's code, so I have no idea where these numbers are from

File: DCA/adapter.py (hunk 15 above)
Comment (maybenotilya) [reply DCA/adapter.py]: Is it ok to fix it with using `os.environ["CUDA_VISIBLE_DEVICES"]` ?

File: DCA/Dockerfile [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+
+RUN apt-get update \
+    && apt-get install --no-install-recommends -y \
+    unzip \
+    wget \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN pip install gdown \
+    && mkdir weights \
+    && gdown "https://drive.google.com/uc?export=download&id=1ZPsIJdEFLDvVSvguOyJj1WiCTANZiYs0" -O ./weights/RURAL_0.4517.pth \
```
Comment (kirill-ivanov-a) [inline DCA/Dockerfile]: It is better to create a separate variable using the ARG command, which will store the path to the weights

File: DCA/adapter.py (hunk 4 above)
Comment (kirill-ivanov-a) [reply DCA/adapter.py:15]: I don't see any fixes

File: DCA/adapter.py (hunk 15 above)
Comment (kirill-ivanov-a) [reply DCA/adapter.py]: Nope

Comment (kirill-ivanov-a) [reply DCA/adapter.py]: It's better to explicitly give the user the ability to switch execution context

File: DCA/adapter.py (hunk 14 above)
Comment (kirill-ivanov-a) [reply DCA/adapter.py]: Not the best argument. You need to understand why they are needed and remove the magic numbers. How can this code be maintained if even the author does not know what these numbers mean?

File: DCA/adapter.py [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+from DCA.utils.my_tools import pre_slide
+from DCA.module.Encoder import Deeplabv2
+
+
+class DcaAdapter(AdapterBase):
+    
+    def __init__(self, image_size=1024, factor=2, mode="Urban"):
+        """
+        Image Size -- size to scale image to (algorithm requires square images)
+        Factor -- algorithm runs segmentation on square patches of image, factor shows how to scale each side of image to make these patches (if factor=n, there will be n^2 patches)
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: The line is too long

File: DCA/adapter.py [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+            use_ppm=True,
+            ppm=dict(
+                num_classes=7,
+                use_aux=False,
+                fc_dim=2048,
+            ),  
+            inchannels=2048,
+            num_classes=7
+        )).to('cuda' if torch.cuda.is_available() else 'cpu')
+        ckpt_path = {"urban" : Path(__file__).parent / "weights" / "URBAN_0.4635.pth",
```
Comment (kirill-ivanov-a) [inline DCA/adapter.py]: Why is the checkpoint name hardcoded? Why can't you just pass the path to it in the constructor?

File: DCA/run.py [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+from adapter_utils import max_power_of_2, get_args
+from adapter import DcaAdapter
+
+
+if __name__ == "__main__":
+    print("Starting work")
+    args = get_args()
+
+    if args.device.lower() == 'cuda' or args.device.lower() == 'gpu':
+        os.environ["CUDA_VISIBLE_DEVICES"] = "0"
```
Comment (kirill-ivanov-a) [inline DCA/run.py]: It's not a good idea to change the environment from Python code. This variable must be set by the user before starting if he wants to work with the GPU

**Review by kirill-ivanov-a — CHANGES_REQUESTED** [review]

File: DCA/run.py [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+
+from pathlib import Path
+from skimage.io import imread
+
+from adapter_utils import get_args
+from adapter import DcaAdapter
+
+
+def main():
+    print("Starting work")
```
Comment (kirill-ivanov-a) [inline DCA/run.py]: [Better use logger](https://docs.python.org/3/library/logging.html)
image

File: DCA/run.py [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+    else:
+        print(f"Failed to recognize device {args.device}")
+        return
+
+    adapter = DcaAdapter(factor=args.factor, model=args.model, device=device)
+
+    input_path = Path(__file__).parent / "input"
+    output_path = Path(__file__).parent / "output"
+
+    for fname in os.listdir(input_path):
```
Comment (kirill-ivanov-a) [inline DCA/run.py]: use pathlib

File: DCA/run.py [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+        print(fname)
+        image = imread(input_path / fname)
+        mask = adapter.process(image)
+        output_name = '.'.join(fname.split('.')[:-1]) if fname.count('.') != 0 else fname
```
Comment (kirill-ivanov-a) [inline DCA/run.py]: Using pathlib you can make this line more understandable (`stem`)

**Review by kirill-ivanov-a — APPROVED** [review]

File: DCA/adapter_base.py [hunk 32]
```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2022, Ilya Syresenkov, Kirill Ivanov and Anastasiia Kornilova
```
Comment (anastasiia-kornilova) [inline DCA/adapter_base.py]: Adapter base is common for all methods, let's move to separate external folder `common` as in Kirill examples

File: DCA/adapter_utils.py [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def get_args():
+    parser = ArgumentParser()
+    parser.add_argument("-f", "--factor", type=int, default=2,
+                        help="Factor shows how images must be scaled to create patches, for factor = n there will be "
+                             "n^2 patches (default: 2)")
+    parser.add_argument("-m", "--model", type=Path, default=Path(__file__, "weights", "Urban.pth"),
+                        help="Pretrained model path (default: weights/Urban.pth)")
+    parser.add_argument("-d", "--device", type=str, default='cuda',
```
Comment (anastasiia-kornilova) [inline DCA/adapter_utils.py]: How do you choose what type of quotes to use? I see strange mix of everything (single/double), let's setup one style and formatter for it

File: README.md [hunk 34]
```diff
@@ -1 +1,2 @@
-# semantic_segmentation_adapters
\ No newline at end of file
+# Semantic Segmentation Adapters
+This folder contains adapters and docker wrappers for aerial images semantic segmentation methods
```
Comment (anastasiia-kornilova) [inline README.md:2]: Please, add here table with supported dockers as here https://github.com/prime-slam/line-detection-association-dockers/tree/main

File: DCA/adapter_utils.py [hunk 35]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2022, Ilya Syresenkov, Kirill Ivanov and Anastasiia Kornilova
```
Comment (anastasiia-kornilova) [inline DCA/adapter_utils.py]: Is this adapter utils? I think it is related to run execution, maybe simple rename it to utils

File: DCA/adapter_base.py [hunk 36]
```diff
@@ -0,0 +1,45 @@
...
+
+    def process(self, image):
+        # image - RGB
+        image = self._transform_image(image)
+        model = self._build_model()
+        raw_predictions = self._process(model, image)
+        predictions = self._postprocess_predictions(raw_predictions)
+        return predictions
+
+    @abstractmethod
```
Comment (anastasiia-kornilova) [inline DCA/adapter_base.py]: docstrings are missed

File: DCA/run.py [hunk 37]
```diff
@@ -0,0 +1,52 @@
...
+from pathlib import Path
+from skimage.io import imread
+
+from adapter_utils import get_args
+from adapter import DcaAdapter
+
+logging.basicConfig(level=logging.INFO)
+
+
+def main():
```
Comment (anastasiia-kornilova) [inline DCA/run.py]: What for you put it to separate function? It doesn't have any sense, let's put things after `if __name__ == "__main__":`

File: DCA/run.py [hunk 38]
```diff
@@ -0,0 +1,52 @@
...
+        device = "cpu"
+    elif args.device.lower() in ("cuda", "gpu"):
+        device = "cuda"
+    else:
+        logging.error(f" Failed to recognize device {args.device}")
+        return
+
+    adapter = DcaAdapter(factor=args.factor, model=args.model, device=device)
+
+    input_dir = Path("/", "DCA", "input")
```
Comment (anastasiia-kornilova) [inline DCA/run.py]: Let's not hardcode, it should be arguments in the script with those defaults

**Review by anastasiia-kornilova — CHANGES_REQUESTED** [review]

**Review by anastasiia-kornilova — APPROVED** [review]

**Pull request merged**